src/lib/core/utils/data-element-constraints.ts

```diff
@@ -462,16 +459,31 @@ export function leastAncestralVariable(
   // The least ancestral node has the most ancestors, so let's count them.
   // And sort by the counts, most-first
   const leastAncestralFirst = sortBy(variables, (variable) => {
-    const ancestorCount = ancestorEntitiesForVariable(variable, entities)
-      .length;
+    const ancestorCount = ancestorEntitiesForEntityId(
+      variable.entityId,
+      entities
+    ).length;
     return -ancestorCount;
   });
 
   // we don't care about ties, because the same-branch constraint means that
   // the two variables will be from the same entity if they have the same number of ancestors
   const variable = leastAncestralFirst[0];
-  if (VariableDescriptor.is(variable)) return variable;
-  else return undefined;
+  if (VariableDescriptor.is(variable)) {
+    // Do a final check that all the input variables are all from ancestor entities of this leaf-most variable's entity.
+    // Thus ensuring that they are indeed on the same branch.
+    const ancestors = ancestorEntitiesForEntityId(variable.entityId, entities);
+
+    if (
+      variables.every(
+        (v) =>
+          isEqual(v, variable) || ancestors.find((e) => e.id === v.entityId)
+      )
+    )
+      return variable;
+    else
+      throw 'Error: variables not expected to be in different branches of entity diagram';
```

Comment:
I just noticed this. You should throw an instance of an Error object, eg `throw new Error(message)`